### PR TITLE
CANParser timing benchmark script

### PR DESCRIPTION
**Pull request by sshane**
didn't see the timing test in https://github.com/commaai/openpilot/issues/27695

runs on demo route:

```python
batman@workstation-shane:~/openpilot/selfdrive/debug$ ./check_can_parser_performance.py 
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  3.10it/s]
6000 CAN packets, 10 runs
322.54 mean ms, 326.21 max ms, 320.80 min ms, 1.61 std ms
0.0538 mean ms / CAN packet
batman@workstation-shane:~/openpilot/selfdrive/debug$ ./check_can_parser_performance.py 
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  3.06it/s]
6000 CAN packets, 10 runs
326.24 mean ms, 347.83 max ms, 320.97 min ms, 8.82 std ms
0.0544 mean ms / CAN packet
batman@workstation-shane:~/openpilot/selfdrive/debug$ ./check_can_parser_performance.py 
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  3.00it/s]
6000 CAN packets, 10 runs
333.05 mean ms, 344.65 max ms, 325.36 min ms, 5.09 std ms
0.0555 mean ms / CAN packet
```

- adds/fixes test_models typing
- script to measure time to parse demo route, using test_model's helpers to reduce duplication

### CODE REVIEW

File: selfdrive/car/tests/test_models.py [hunk 1]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
         can_invalid_cnt += not CS.canValid
 
     self.assertEqual(can_invalid_cnt, 0)
+    self.assertInterfaceTiming((time.process_time_ns() - start_time) * 1e-6)
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (1000, 3000)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:188]:
```suggestion
    total_time_thresh = (1000, 5000)
```

File: selfdrive/car/tests/test_models.py [hunk 2]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
...
+    per_packet_thresh = (0.2, 0.5)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:189]:
```suggestion
    per_packet_thresh = (0.2, 0.6)
```

File: selfdrive/car/tests/test_models.py [hunk 3]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
...
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (1000, 5000)
+    per_packet_thresh = (0.2, 0.6)
+
+    avg_time_msg = total_time_ms / len(self.can_msgs)
+    for t, thresh in ((total_time_ms, total_time_thresh), (avg_time_msg, per_packet_thresh)):
+      self.assertLess(t, thresh[1])
+      self.assertGreater(t, thresh[0], "Performance seems to have improved, update test refs.")
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:194]:
```suggestion
    #for t, thresh in ((total_time_ms, total_time_thresh), (avg_time_msg, per_packet_thresh)):
    #  self.assertLess(t, thresh[1])
    #  self.assertGreater(t, thresh[0], "Performance seems to have improved, update test refs.")
```

File: selfdrive/car/tests/test_models.py [hunk 4]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
         can_invalid_cnt += not CS.canValid
 
     self.assertEqual(can_invalid_cnt, 0)
+    self.assertInterfaceTiming((time.process_time_ns() - start_time) * 1e-6)
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (1000, 5000)
+    per_packet_thresh = (0.2, 0.6)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:189]:
```suggestion
    per_packet_thresh = (0.15, 1.0)
```

File: selfdrive/car/tests/test_models.py [hunk 5]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
         can_invalid_cnt += not CS.canValid
 
     self.assertEqual(can_invalid_cnt, 0)
+    self.assertInterfaceTiming((time.process_time_ns() - start_time) * 1e-6)
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (1000, 5000)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:188]:
```suggestion
    total_time_thresh = (500, 6000)
```

File: selfdrive/car/tests/test_models.py (hunk 4 above)
Comment (sshane) [reply selfdrive/car/tests/test_models.py:189]: seen up to 0.7

File: selfdrive/car/tests/test_models.py [hunk 6]
```diff
@@ -179,6 +181,19 @@ def test_car_interface(self):
...
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (500, 7000)
+    per_packet_thresh = (0.15, 1.0)
+
+    avg_time_msg = total_time_ms / len(self.can_msgs)
+    #for t, thresh in ((total_time_ms, total_time_thresh), (avg_time_msg, per_packet_thresh)):
+    #  self.assertLess(t, thresh[1])
+    #  self.assertGreater(t, thresh[0], "Performance seems to have improved, update test refs.")
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:194]:
```suggestion
    for t, thresh in ((total_time_ms, total_time_thresh), (avg_time_msg, per_packet_thresh)):
      self.assertLess(t, thresh[1])
      self.assertGreater(t, thresh[0], "Performance seems to have improved, update test refs.")
```

File: .github/workflows/selfdrive_tests.yaml [hunk 7]
```diff
@@ -369,7 +369,7 @@ jobs:
     - name: Test car models
       timeout-minutes: 25
       run: |
-        ${{ env.RUN }} "coverage run -m pytest selfdrive/car/tests/test_models.py && \
+        ${{ env.RUN }} "coverage run -m pytest -s selfdrive/car/tests/test_models.py && \
```
Comment (sshane) [inline .github/workflows/selfdrive_tests.yaml:372]:
```suggestion
        ${{ env.RUN }} "coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py || coverage run -m pytest -s selfdrive/car/tests/test_models.py && \
```

File: .github/workflows/selfdrive_tests.yaml [hunk 8]
```diff
@@ -371,9 +371,9 @@ jobs:
     - name: Build openpilot
       run: ${{ env.RUN }} "scons -j$(nproc)"
     - name: Test car models
-      timeout-minutes: 25
+      timeout-minutes: 999
       run: |
-        ${{ env.RUN }} "coverage run -m pytest selfdrive/car/tests/test_models.py && \
+        ${{ env.RUN }} "coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py ;  coverage run -m selfdrive/car/tests/test_models.py && \
```
Comment (sshane) [inline .github/workflows/selfdrive_tests.yaml:376]:
```suggestion
        ${{ env.RUN }} "selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py && \
```

File: .github/workflows/selfdrive_tests.yaml [hunk 9]
```diff
@@ -371,9 +371,9 @@ jobs:
     - name: Build openpilot
       run: ${{ env.RUN }} "scons -j$(nproc)"
     - name: Test car models
-      timeout-minutes: 25
+      timeout-minutes: 999
       run: |
-        ${{ env.RUN }} "coverage run -m pytest selfdrive/car/tests/test_models.py && \
+        ${{ env.RUN }} "selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py ;  selfdrive/car/tests/test_models.py && \
```
Comment (sshane) [inline .github/workflows/selfdrive_tests.yaml:376]:
```suggestion
        ${{ env.RUN }} "./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py ;  ./selfdrive/car/tests/test_models.py && \
```

File: selfdrive/car/tests/test_models.py [hunk 10]
```diff
@@ -179,6 +183,21 @@ def test_car_interface(self):
         can_invalid_cnt += not CS.canValid
 
     self.assertEqual(can_invalid_cnt, 0)
+    self.assertInterfaceTiming((time.process_time_ns() - start_time) * 1e-6)
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (500, 7000)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:190]:
```suggestion
    total_time_thresh = (100, 7000)
```

File: selfdrive/car/tests/test_models.py [hunk 11]
```diff
@@ -179,6 +183,21 @@ def test_car_interface(self):
         can_invalid_cnt += not CS.canValid
 
     self.assertEqual(can_invalid_cnt, 0)
+    self.assertInterfaceTiming((time.process_time_ns() - start_time) * 1e-6)
+
+  def assertInterfaceTiming(self, total_time_ms):
+    """Asserts sane timing"""
+    total_time_thresh = (100, 7000)
+    per_packet_thresh = (0.15, 1.0)
```
Comment (sshane) [inline selfdrive/car/tests/test_models.py:191]:
```suggestion
    per_packet_thresh = (0.10, 0.9)
```

**Comment by sshane** [comment]
Out of 1500 test runs:

```python
print(np.mean(all_avg_runs), np.std(all_avg_runs), min(all_avg_runs), max(all_avg_runs))
0.3691224778989577 0.10412673168843738 0.1797995685363425 0.6925979475484606
```

**Comment by sshane** [comment]
ran for 1 hour in CI with no failures, total time has to be wide because some routes have far less CAN traffic than others.

**Comment by adeebshihadeh** [comment]
This test won't work, even in CI, since it relies on the machine it's run on. The best we can do here is make this a benchmarking script.

**Comment by sshane** [comment]
@adeebshihadeh something like this?

**Comment by adeebshihadeh** [comment]
Yes

**Pull request merged**